### PR TITLE
Use column statistics to generate the physical plan in cuDF-Polars

### PR DESCRIPTION
**Pull request by rjzamora**
## Description
Closes https://github.com/rapidsai/cudf/issues/19391

Allows cuDF-Polars to (optionally) use column statistics to generate the physical plan. Local benchmarking suggests that the physical plan is usually **conservative** with the new option enabled, because the statistics do **not** capture filters. We also don't try to inject `Repartition` operations.

Since statistics-based planning is disabled by default, we probably don't need to further optimize the `update_column_stats` logic in this PR. We can probably focus on the general design and configuration API. With that said, we probably want to improve things a bit more before enabling statistics-based planning by default.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf_polars/cudf_polars/experimental/benchmarks/utils.py [hunk 1]
```diff
@@ -375,15 +382,15 @@ def print_query_plan(
...
+        if args.explain:
             print(f"\nQuery {q_id} - Physical plan\n")
             print(q.show_graph(engine="streaming", plan_stage="physical"))
     elif CUDF_POLARS_AVAILABLE:
         assert isinstance(engine, pl.GPUEngine)
         if args.explain_logical:
             print(f"\nQuery {q_id} - Logical plan\n")
             print(explain_query(q, engine, physical=False))
-        elif args.explain:
+        if args.explain:
```
Comment (rjzamora) [inline python/cudf_polars/cudf_polars/experimental/benchmarks/utils.py:393]: I can move this to a separate PR, but it's sometimes useful to print both the `--explain` and `--explain-logical` result in the same run.

File: python/cudf_polars/cudf_polars/utils/config.py [hunk 2]
```diff
@@ -289,6 +290,61 @@ def default_blocksize(scheduler: str) -> int:
     return min(max(blocksize, 1_000_000_000), 10_000_000_000)
 
 
+@dataclasses.dataclass(frozen=True)
+class StatsPlanningOptions:
```
Comment (rjzamora) [inline python/cudf_polars/cudf_polars/utils/config.py:294]: These new "nested" config options are probably the most important part of this PR.

File: python/cudf_polars/cudf_polars/utils/config.py [hunk 3]
```diff
@@ -289,6 +290,81 @@ def default_blocksize(scheduler: str) -> int:
...
+    """
+    Configuration for statistics-based query planning.
+
+    These options can be configured via environment variables
+    with the prefix ``CUDF_POLARS__STATISTICS_PLANNING__``.
+
+    Parameters
+    ----------
+    io_partitioning
```
Comment (TomAugspurger) [inline python/cudf_polars/cudf_polars/utils/config.py]: For consistency, could we have all the boolean options start with `use_`?

And is the long-term goal to have `reduction_planning` enabled by default? It'd be nice if they were all true or all false by default.

File: python/cudf_polars/cudf_polars/utils/config.py [hunk 4]
```diff
@@ -289,6 +290,81 @@ def default_blocksize(scheduler: str) -> int:
...
+        the output-partition count for reduction operations
+        like `Distinct`, `GroupBy`, and `Select(unique)`.
+        Default is False.
+    use_join_heuristics
+        Whether to use join heuristics to estimate row-count
+        and unique-count statistics. Default is True.
+        These statistics may only be collected when they are
+        actually needed for query planning.
+    use_sampling
+        Whether to sample real data to estimate unique-value
```
Comment (TomAugspurger) [inline python/cudf_polars/cudf_polars/utils/config.py:321]: How many of these options are only relevant for parquet sources? If any, let's consider documenting that.

File: python/cudf_polars/cudf_polars/utils/config.py [hunk 5]
```diff
@@ -289,6 +290,81 @@ def default_blocksize(scheduler: str) -> int:
     return min(max(blocksize, 1_000_000_000), 10_000_000_000)
 
 
+@dataclasses.dataclass(frozen=True)
+class StatsPlanningOptions:
+    """
+    Configuration for statistics-based query planning.
+
+    These options can be configured via environment variables
+    with the prefix ``CUDF_POLARS__STATISTICS_PLANNING__``.
```
Comment (TomAugspurger) [inline python/cudf_polars/cudf_polars/utils/config.py]: In other frameworks like pydantic-settings, nested configuration is typically handled by putting the full path to value being configured, separated by `__`. So this would be something like


```
CUDF_POLARS__EXECUTOR__STATS_PLANNING_OPTIONS__IO_PARTITIONING=0
```

It's a bit verbose, but unambiguous and the consistency is nice to have.

I might recommend renaming the attribute to `statistics` (or `stats`), so it'd be

```
CUDF_POLARS__EXECUTOR__STATISTICS__IO_PARTITIONING
```

File: python/cudf_polars/tests/experimental/test_stats.py [hunk 6]
```diff
@@ -407,3 +410,81 @@ def test_base_stats_join_key_info(engine):
...
+            pl.col("amount").sum().alias("total_amount"),
+            pl.col("order_id").count().alias("order_count"),
+            pl.col("customer_name").first().alias("name"),
+            pl.col("region").first().alias("region"),
+        ]
+    )
+    q = q_gb.sort("customer_id")
+
+    # Verify the query runs correctly
+    assert_gpu_result_equal(q, engine=engine)
```
Comment (TomAugspurger) [inline python/cudf_polars/tests/experimental/test_stats.py:490]: What guarantees do we have that this actually did exercise all the choices made by the various statistics options?

I think if we have other tests that actually get the IR node, collects the statistics, and makes assertions about them, then having this just as a basic correctness test is good. If we don't have those tests, then perhaps we could add them or update this test to collect the stats and make some assertions, depending on which options are enabled.

File: python/cudf_polars/cudf_polars/utils/config.py (hunk 3 above)
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/utils/config.py]: >For consistency, could we have all the boolean options start with use_?

Yeah, that works for me.

>And is the long-term goal to have reduction_planning enabled by default? It'd be nice if they were all true or all false by default.

Yes. The problem is that we (probably) see a slight performance hit in our PDS-H benchmarks when we turn it on for now. This is because we actually use tree reductions in many places by default, and when we start gathering statistics, some of those reductions are switched to shuffles. In some cases, we are just being a bit aggressive without statistics. In other cases, we are okay with the tree reduction because there are filters that drop 99% of the data (and it's nearly impossible to know this by inspecting the plan and the statistics).

It seems reasonable to get these changes in before taking the time to address these regressions (with `unique_fraction` adjustments and/or statistics improvements). There are likely to be users who will benefit from turning this option on, because it should make those `unique_fraction` configs unnecessary.

File: python/cudf_polars/cudf_polars/utils/config.py (hunk 4 above)
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/utils/config.py:321]: I added some notes about the options that only apply to Parquet and LazyFrame data.

File: python/cudf_polars/cudf_polars/utils/config.py (hunk 5 above)
Comment (rjzamora) [reply python/cudf_polars/cudf_polars/utils/config.py]: Heh - This was some auto-complete stuff that got copied everywhere :)

I changed the name from `stats_planning_options` to `stats_planning` to be slightly-more concise. I also added the missing `__EXECUTOR__` string.

File: python/cudf_polars/tests/experimental/test_stats.py (hunk 6 above)
Comment (rjzamora) [reply python/cudf_polars/tests/experimental/test_stats.py:490]: Yeah, the purpose of these tests are mostly to get code coverage for several edge cases (mostly where we are using empty/default `ColumnStat` objects in our logic). We do most of the statistics testing in `test_explain.py` because "white-box" testing of the `StatsCollector` object turned out to be an extremely painful (i.e. verbose, fragile, and unreadable) way to build test coverage.
